### PR TITLE
Add mypy for utils

### PR DESCRIPTION
**Pull request by AmedeoBiolatti**
## Motivation

mypy coverage for utils module, as for https://github.com/sgl-project/sglang-jax/issues/458

## Modifications

- added `"python/sgl_jax/srt/utils/*.py"` mypy to `[tool.mypy]`
- solved minor typing annotation issues 

## Accuracy Tests

no change

## Benchmarking and Profiling

no change

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/utils/jax_utils.py [hunk 1]
```diff
@@ -198,11 +197,14 @@ def _get_available_memory_distributed(a):
 
 def device_array(*data, sharding=None, **kwargs) -> jax.Array:
     if sharding is None:
-        return jax.device_put(*data, device=sharding, **kwargs)
+        return jax.device_put(data, device=sharding, **kwargs)
```
Comment (Rodrian7) [inline python/sgl_jax/srt/utils/jax_utils.py:200]: jax.device_put(*data, ...) → jax.device_put(data, ...) — this changes runtime behavior, not just type annotations. Was this intentional?

File: python/sgl_jax/srt/utils/jax_utils.py [hunk 2]
```diff
@@ -198,11 +197,14 @@ def _get_available_memory_distributed(a):
...
 
     def _to_device(arr):
+        def fn(idx, a=arr):
+            return a[idx]
+
         arr = np.asarray(arr)
-        return jax.make_array_from_callback(arr.shape, sharding, lambda idx, a=arr: a[idx])
+        return jax.make_array_from_callback(arr.shape, sharding, fn)
```
Comment (Rodrian7) [inline python/sgl_jax/srt/utils/jax_utils.py:208]: In _to_device, fn is defined before arr = np.asarray(arr), so the callback receives the original JAX array instead of the converted numpy one. Differs from the original behavior.

File: python/sgl_jax/srt/utils/weight_utils.py [hunk 3]
```diff
@@ -1904,7 +1921,7 @@ def load_weights_from_safetensors(
 
                 infos = weight_info[hf_key]
 
-                if isinstance(mapping, str | list):
+                if not isinstance(mapping, WeightMapping):
```
Comment (Rodrian7) [inline python/sgl_jax/srt/utils/weight_utils.py]: This changes behavior for tuple mappings — previously they were not wrapped into WeightMapping, now they are. Was this intentional?

File: python/sgl_jax/srt/utils/jax_utils.py (hunk 1 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/jax_utils.py:200]: `jax.device_put(*data, device=sharding, **kwargs)` doesn't seem to be compatible in general with `jax.device_put`, since there are only two positional arguments and the second one is already passed with `device=sharding`, so `device_array(x, y)` could raise an error.
 (https://docs.jax.dev/en/latest/_autosummary/jax.device_put.html)

File: python/sgl_jax/srt/utils/weight_utils.py [hunk 4]
```diff
@@ -64,8 +65,8 @@ def _reinterpret_dtype_if_needed(data: np.ndarray, target_dtype: jnp.dtype) -> n
 
 @dataclass
 class WeightMapping:
-    target_path: str | list[str]
-    sharding: tuple | None = None
+    target_path: str | list[str] | tuple[str]
+    sharding: tuple[Any] | tuple[Any, Any] | None = None
```
Comment (Rodrian7) [inline python/sgl_jax/srt/utils/weight_utils.py]: sharding: tuple[Any] | tuple[Any, Any] doesn't cover 3-element tuples like ("expert", None, None) — should be tuple[Any, ...].
What's the motivation for adding tuple[str] to target_path?

File: python/sgl_jax/srt/utils/jax_utils.py (hunk 1 above)
Comment (Rodrian7) [reply python/sgl_jax/srt/utils/jax_utils.py:200]: Good point. But with data, single-arg callers like device_array(np.array(...)) now get a tuple back instead of a single array — does that need handling?

Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/jax_utils.py:200]: We could move the definition to `def device_array(data, sharding=None, **kwargs) -> jax.Array:`, since it is always used with one positional argument and it is more consistent with jax

File: python/sgl_jax/srt/utils/jax_utils.py (hunk 2 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/jax_utils.py:208]: I missed the fact that defaults are evaluated at def time in python, good to know. I will fix this

File: python/sgl_jax/srt/utils/weight_utils.py (hunk 3 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/weight_utils.py]: The code here is already expecting `str | list | WeightMapping` and a `tuple` would raise an error in the old version (there is `mapping.concat_axis` a couple of lines below), but you're right, we should keep the behaviour and make it more explicity. I will check type and raise an error in case of a tuple

File: python/sgl_jax/srt/utils/weight_utils.py (hunk 4 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/weight_utils.py]: - fixed sharding
- target_path is related to the previous comment about tuple->WeightMapping, I added the tuple type to make it compatible with tuple conversion to WeightMapping, but now there there are more checks I removed it

File: python/sgl_jax/srt/utils/weight_utils.py (hunk 3 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/weight_utils.py]: done, now I added a stricter check on the mapping type

File: python/sgl_jax/srt/utils/jax_utils.py (hunk 2 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/jax_utils.py:208]: fixed

File: python/sgl_jax/srt/utils/jax_utils.py (hunk 1 above)
Comment (AmedeoBiolatti) [reply python/sgl_jax/srt/utils/jax_utils.py:200]: fixed

**Review by Rodrian7 — APPROVED** [review]

**Comment by cjx0709** [comment]
Temporarily closing and reopening to retrigger PR CI on the latest head commit.
